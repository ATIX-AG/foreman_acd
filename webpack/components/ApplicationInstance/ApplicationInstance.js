import React, { useState } from 'react'
import PropTypes from 'prop-types';
import {
  Icon,
  Button,
} from 'patternfly-react';
import * as resolve from 'table-resolver';
import ForemanModal from 'foremanReact/components/ForemanModal';
import {
  sprintf,
  translate as __
} from 'foremanReact/common/I18n';
import Select from 'foremanReact/components/common/forms/Select';
import ParameterSelection from '../ParameterSelection';
import AddTableEntry from '../common/AddTableEntry';
import DeleteTableEntry from '../common/DeleteTableEntry';
import RailsData from '../common/RailsData';
import EasyHeaderFormatter from '../common/EasyHeaderFormatter';
import AppDefinitionSelector from './components/AppDefinitionSelector';
import ServiceCounter from './components/ServiceCounter';
import { arrayToObject } from '../../helper';

import {
  Table,
  FormControl,
  inlineEditFormatterFactory,
} from 'patternfly-react';

import {
  PARAMETER_SELECTION_PARAM_TYPE_FOREMAN,
  PARAMETER_SELECTION_PARAM_TYPE_ANSIBLE,
} from '../ParameterSelection/ParameterSelectionConstants';

class ApplicationInstance extends React.Component {

  constructor(props) {
    super(props);
  }

  isEditing({rowData}) {
    return (rowData.backup !== undefined);
  }

  addTableEntryAllowed() {
    return this.props.editMode || this.props.appDefinition.id == ''
  }

  validateParameters() {
    let result = true;
    let msg = "";

    this.props.hosts.forEach(h => {
      if (h.foremanParameters.map(e => e.value).filter(i => i == "").length > 0) {
        result = false;

        if (msg == "") {
          msg += __("For some hosts the values for some parameters are missing. Check the values for these hosts:\n");
        }
        msg += "- "+ h.hostname +"\n";
      }
    });

    const invalidMinServices = this.props.services.filter(s => (Number(s.minCount) != 0) && (s.currentCount < s.minCount));
    const invalidMaxServices = this.props.services.filter(s => (Number(s.maxCount) != 0) && (s.currentCount > s.maxCount));

    console.log(invalidMinServices);

    if (invalidMinServices.length > 0 || invalidMaxServices.length > 0) {
      result = false;

      if (msg != "") {
        msg += "\n";
      }

      msg += __("Unachieved service counts:");
      msg += "\n";

      invalidMinServices.map(s => { msg += sprintf(
        __(`- service ${s.name} expects at ${s.minCount} least configured hosts\n`)
      )});

      invalidMaxServices.map(s => { msg += sprintf(
        __(`- service ${s.name} expects no more than ${s.axCount} configured hosts\n`)
      )});
    }

    return {
      validateResult: result,
      validateMsg: msg
    }
  }

  componentDidMount() {
    const {
      data: { mode, appDefinition, hosts, ansibleVarsAll, appDefinitionUrl },
      initApplicationInstance,
      addApplicationInstanceHost,
      deleteApplicationInstanceHost,
      activateEditApplicationInstanceHost,
      changeEditApplicationInstanceHost,
      openForemanParameterSelectionModal,
      openAnsibleParameterSelectionModal,
      loadApplicationDefinition,
    } = this.props;

    if (mode === 'editInstance') {
      loadApplicationDefinition(appDefinition.id, { url: appDefinitionUrl });
    }

    const inlineEditButtonsFormatter = inlineEditFormatterFactory({
      isEditing: additionalData => this.props.editMode,
      renderValue: (value, additionalData) => (
        <td style={{ padding: '2px' }}>
          <Button
            bsStyle="default"
            onClick={() => activateEditApplicationInstanceHost(additionalData)}
          >
            <Icon type="pf" name="edit" title={__("edit entry")} />
          </Button>
          &nbsp;
          <Button
            bsStyle="default"
            onClick={() => openForemanParameterSelectionModal(additionalData)}
          >
            <Icon type="pf" name="settings" title={__("change parameters")} />
          </Button>
          &nbsp;
          <Button
            bsStyle="default"
            onClick={() => openAnsibleParameterSelectionModal(additionalData)}
          >
            <span title={__("change ansible variables")}>A</span>
          </Button>
          &nbsp;
          <DeleteTableEntry
            hidden={false}
            disabled={false}
            onDeleteTableEntry={deleteApplicationInstanceHost}
            additionalData={additionalData}
          />
        </td>
      ),
      renderEdit: (value, additionalData) => (
        <td style={{ padding: '2px' }}>
          <Button bsStyle="default" disabled>
            <Icon type="pf" name={__("edit")} />
          </Button>
          &nbsp;
          <Button bsStyle="default" disabled>
            <Icon type="pf" name={__("settings")} />
          </Button>
          &nbsp;
          <Button bsStyle="default" disabled>
            <span>A</span>
          </Button>
          &nbsp;
          <DeleteTableEntry
            hidden={false}
            disabled={true}
            onDeleteTableEntry={deleteApplicationInstanceHost}
            additionalData={additionalData}
          />
        </td>
      )
    });
    this.inlineEditButtonsFormatter = inlineEditButtonsFormatter;

    this.headerFormatter = EasyHeaderFormatter;

    const inlineEditFormatterImpl = {
      renderValue: (value, additionalData) => (
        <td>
          <span className="static">{value}</span>
        </td>
      ),
      renderEditText: (value, additionalData, subtype='text') => (
        <td className="editing">
          <FormControl
            type={subtype}
            defaultValue={value}
            onBlur={e => changeEditApplicationInstanceHost(e.target.value, additionalData) }
          />
        </td>
      ),
      renderEditSelect: (value, additionalData, options) => (
        <td className="editing">
          <Select
            value={value.toString()}
            onChange={e => changeEditApplicationInstanceHost(e.target.value, additionalData) }
            options={options}
            allowClear
            key="key"
          />
        </td>
      )
    };

    const inlineEditFormatter = inlineEditFormatterFactory({
      isEditing: additionalData => this.isEditing(additionalData),
      renderValue: (value, additionalData) => {
        let prettyValue = value;
        if (additionalData.property == 'service') {
          const serviceList = arrayToObject(this.props.services, "id", "name");
          prettyValue = serviceList[value];
        }
        return inlineEditFormatterImpl.renderValue(prettyValue, additionalData)
      },
      renderEdit: (value, additionalData) => {
        let prettyValue = value;
        if (additionalData.property == 'service') {
          const availableServices = this.props.services.filter(service => ((Number(service['maxCount']) == 0) || (service['currentCount'] < service['maxCount'])));
          const serviceList = arrayToObject(availableServices, "id", "name");

          if (additionalData.rowData.newEntry === true) {
            return inlineEditFormatterImpl.renderEditSelect(value, additionalData, serviceList);
          }
          prettyValue = serviceList[value];
          return inlineEditFormatterImpl.renderValue(prettyValue, additionalData)
        }
	if (additionalData.property == 'hostname') {
	  if (additionalData.rowData.newEntry === true) {
            return inlineEditFormatterImpl.renderEditText(value, additionalData);
          }
          return inlineEditFormatterImpl.renderValue(prettyValue, additionalData)
        }
        return inlineEditFormatterImpl.renderEditText(prettyValue, additionalData);
      }
    });
    this.inlineEditFormatter = inlineEditFormatter;

    initApplicationInstance(
      appDefinition,
      hosts,
      ansibleVarsAll,
      this.headerFormatter,
      this.inlineEditFormatter,
      this.inlineEditButtonsFormatter,
    );
  };

  render() {
    const {
      data: { mode, applications, organization, location, foremanDataUrl, appDefinitionUrl },
      appDefinition,
      services,
      hosts,
      columns,
      addApplicationInstanceHost,
      confirmEditApplicationInstanceHost,
      cancelEditApplicationInstanceHost,
      closeForemanParameterSelectionModal,
      openAnsibleParameterSelectionModal,
      closeAnsibleParameterSelectionModal,
      changeParameterSelectionMode,
      loadApplicationDefinition,
    } = this.props;

    let { validateResult, validateMsg } = this.validateParameters();

    if (validateResult == false) {
      $('input[type="submit"][name="commit"]').attr("disabled", true);
    } else {
      $('input[type="submit"][name="commit"]').attr("disabled", false);
    }

    return (
      <span>
        <div class="service-counter">
          <ServiceCounter
            title="Service counts"
            serviceList={ services }
            hostList={ hosts }
          />
        </div>
        <div>
          <AppDefinitionSelector
            label="Application Definition"
            editable={ mode == 'newInstance' }
            viewText={ appDefinition.name }
            options={ applications }
            onChange={ loadApplicationDefinition }
            selectValue={ appDefinition.id.toString() }
            additionalData={{url: appDefinitionUrl}}
          />
          {appDefinition.id == '' ? (
          <p style={{ paddingTop: 25 }}>
            <pre>{ "App Definition can't be blank" }</pre>
          </p>
        ) : (<div></div>)}
        </div>
        <div className="form-group">
          <AddTableEntry
             hidden={ false }
             disabled={ this.addTableEntryAllowed() }
             onAddTableEntry={ addApplicationInstanceHost }
          />
          <Table.PfProvider
            striped
            bordered
            hover
            dataTable
            inlineEdit
            columns={columns}
            components={{
              body: {
                row: Table.InlineEditRow,
                cell: cellProps => cellProps.children
              }
            }}
          >
            <Table.Header headerRows={resolve.headerRows({ columns })} />
            <Table.Body
              rows={hosts}
              rowKey="id"
              onRow={(rowData, { rowIndex }) => ({
                role: 'row',
                isEditing: () => this.isEditing({ rowData }),
                onCancel: () => cancelEditApplicationInstanceHost({ rowData, rowIndex }),
                onConfirm: () => confirmEditApplicationInstanceHost({ rowData, rowIndex }),
                last: rowIndex === services.length - 1
              })}
            />
          </Table.PfProvider>
          <AddTableEntry
            hidden={ false }
            disabled={ this.addTableEntryAllowed() }
            onAddTableEntry={ addApplicationInstanceHost }
          />
          <span style={{ marginLeft: 30 }}>
            Ansible group vars 'all':
            <Button
              style={{ marginLeft: 10 }}
              bsStyle="default"
              disabled={ this.props.editMode }
              onClick={() => openAnsibleParameterSelectionModal({
                isAllGroup: true
              })}
            >
              <span title={__("change ansible variables for 'all'")}>A</span>
            </Button>
          </span>
        </div>
        <div>
          <ForemanModal
            id="AppInstanceForemanParamSelection"
            dialogClassName="param_selection_modal"
            title={__("Foreman Parameter specification for Application Instance")}
          >
            <ForemanModal.Header closeButton={false}>
              Parameter specification
            </ForemanModal.Header>
            {this.props.parametersData ? (
              <ParameterSelection
                editModeCallback={ (hide) => changeParameterSelectionMode({ mode: hide })}
                paramType={ PARAMETER_SELECTION_PARAM_TYPE_FOREMAN }
                location={ location }
                organization={ organization }
                paramDataUrl= { foremanDataUrl }
                data={ this.props.parametersData }
              />
            ) : (<span>Empty</span>)
            }
            <ForemanModal.Footer>
              <div>
                <Button bsStyle="primary" disabled={this.props.paramEditMode} onClick={() => closeForemanParameterSelectionModal({ mode: 'save' })}>{__("Save")}</Button>
                <Button bsStyle="default" disabled={this.props.paramEditMode} onClick={() => closeForemanParameterSelectionModal({ mode: 'cancel' })}>{__("Cancel")}</Button>
              </div>
            </ForemanModal.Footer>
          </ForemanModal>
        </div>
        <div>
          <ForemanModal
            id="AppInstanceAnsibleParamSelection"
            dialogClassName="param_selection_modal"
            title={__("Ansible group variables for Application Instance")}
          >
            <ForemanModal.Header closeButton={false}>
              Parameter specification
            </ForemanModal.Header>
            {this.props.parametersData ? (
              <ParameterSelection
                editModeCallback={ (hide) => changeParameterSelectionMode({ mode: hide })}
                paramType={ PARAMETER_SELECTION_PARAM_TYPE_ANSIBLE }
                location={ location }
                organization={ organization }
                data={ this.props.parametersData }
              />
            ) : (<span>Empty</span>)
            }
            <ForemanModal.Footer>
              <div>
                <Button bsStyle="primary" disabled={this.props.paramEditMode} onClick={() => closeAnsibleParameterSelectionModal({ mode: 'save' })}>{__("Save")}</Button>
                <Button bsStyle="default" disabled={this.props.paramEditMode} onClick={() => closeAnsibleParameterSelectionModal({ mode: 'cancel' })}>{__("Cancel")}</Button>
              </div>
            </ForemanModal.Footer>
          </ForemanModal>
        </div>
        {validateResult == false ? (
          <p style={{ paddingTop: 25 }}>
            <pre>{ validateMsg }</pre>
          </p>
        ) : (<div></div>)}
        <RailsData
          key='applications_instance'
          view='app_instance'
          parameter='hosts'
          value={JSON.stringify(this.props.hosts)}
        />
        <RailsData
          key='applications_instance'
          view='app_instance'
          parameter='ansible_vars_all'
          value={JSON.stringify(this.props.ansibleVarsAll)}
        />
      </span>
    )};
}

ApplicationInstance.defaultProps = {
  error: {},
  appDefinition: { "id": '', "name": '' },
  editMode: false,
  services: [],
  hosts: [],
  ansibleVarsAll: [],
  parametersData: {},
  columns: [],
  editParamsOfRowId: null,
  paramEditMode: false,
}

ApplicationInstance.propTypes = {
  initApplicationInstance: PropTypes.func,
  editMode: PropTypes.bool.isRequired,
  services: PropTypes.array,
  appDefinition: PropTypes.object,
  columns: PropTypes.array,
  hosts: PropTypes.array,
  ansibleVarsAll: PropTypes.array,
  loadApplicationDefinition: PropTypes.func,
  addApplicationInstanceHost: PropTypes.func,
  deleteApplicationInstanceHost: PropTypes.func,
  activateEditApplicationInstanceHost: PropTypes.func,
  confirmEditApplicationInstanceHost: PropTypes.func,
  cancelEditApplicationInstanceHost: PropTypes.func,
  changeEditApplicationInstanceHost: PropTypes.func,
  openForemanParameterSelectionModal: PropTypes.func,
  closeForemanParameterSelectionModal: PropTypes.func,
  openAnsibleParameterSelectionModal: PropTypes.func,
  closeAnsibleParameterSelectionModal: PropTypes.func,
  changeParameterSelectionMode: PropTypes.func,
  parametersData: PropTypes.object,
  paramEditMode: PropTypes.bool,
};

export default ApplicationInstance;
