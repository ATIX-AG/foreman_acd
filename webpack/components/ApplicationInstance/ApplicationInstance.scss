@import '~@theforeman/vendor/scss/variables';

.service-counter {
  position: absolute;
  right: 200px;
  top: 220px;
}

.service-counter-title {
  font-size: 14px;
}
