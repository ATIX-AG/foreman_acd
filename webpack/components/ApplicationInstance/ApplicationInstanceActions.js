import React from 'react';
import api from 'foremanReact/API';

import {
  actionHeaderCellFormatter,
} from 'patternfly-react';

import {
  propsToSnakeCase,
  propsToCamelCase,
} from 'foremanReact/common/helpers';

import {
  APPLICATION_INSTANCE_INIT,
  APPLICATION_INSTANCE_HOST_DELETE,
  APPLICATION_INSTANCE_HOST_ADD,
  APPLICATION_INSTANCE_HOST_EDIT_ACTIVATE,
  APPLICATION_INSTANCE_HOST_EDIT_CONFIRM,
  APPLICATION_INSTANCE_HOST_EDIT_CHANGE,
  APPLICATION_INSTANCE_HOST_EDIT_CANCEL,
  APPLICATION_INSTANCE_PARAMETER_SELECTION_MODAL_OPEN,
  APPLICATION_INSTANCE_PARAMETER_SELECTION_MODAL_CLOSE,
  APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_REQUEST,
  APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_SUCCESS,
  APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_FAILURE,
} from './ApplicationInstanceConstants';

export const initApplicationInstance = (
  appDefinition,
  hosts,
  headerFormatter,
  inlineEditFormatter,
  inlineEditButtonsFormatter,
) => dispatch => {
  const initialState = {};

  initialState.columns = [
    {
      property: 'hostname',
      header: {
        label: 'Hostname',
        formatters: [headerFormatter],
        props: {
          index: 0,
          style: {
            width: '20%'
          }
        },
      },
      cell: {
        formatters: [inlineEditFormatter]
      }
    },
    {
      property: 'description',
      header: {
        label: 'Description',
        formatters: [headerFormatter],
        props: {
          index: 1,
          style: {
            width: '20%'
          }
        },
      },
      cell: {
        formatters: [inlineEditFormatter]
      }
    },
    {
      property: 'service',
      header: {
        label: 'Service',
        formatters: [headerFormatter],
        props: {
          index: 2,
          style: {
            width: '20%'
          }
        },
      },
      cell: {
        formatters: [inlineEditFormatter]
      }
    },
    {
      property: 'actions',
      header: {
        label: 'Actions',
        formatters: [actionHeaderCellFormatter],
        props: {
          index: 4,
          style: {
            width: '20%'
          }
        },
      },
      cell: {
        formatters: [inlineEditButtonsFormatter]
      }
    }
  ];

  initialState.appDefinition = appDefinition;
  initialState.hosts = hosts;

  dispatch({
    type: APPLICATION_INSTANCE_INIT,
    payload: initialState,
  });
};

const errorHandler = (msg, err) => {
  const error = {
    errorMsg: 'Failed to fetch data from server.',
    statusText: err,
  };
  return { type: msg, payload: { error } };
};

export const loadApplicationDefinition = (
  applicationDefinitionId,
  additionalData,
) => dispatch => {
  dispatch({ type: APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_REQUEST });

  const realUrl = additionalData.url.replace("__id__", applicationDefinitionId);

  return api
    .get(realUrl, {}, {})
    .then(({ data }) =>
      dispatch({
        type: APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_SUCCESS,
        payload: data,
      })
    )
    .catch(error => dispatch(errorHandler(APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_FAILURE, error)));
};

export const addApplicationInstanceHost = (additionalData) => ({
  type: APPLICATION_INSTANCE_HOST_ADD,
  payload: {
    ...additionalData,
  },
});

export const deleteApplicationInstanceHost = (additionalData) => ({
  type: APPLICATION_INSTANCE_HOST_DELETE,
  payload: {
    ...additionalData,
  },
});

export const activateEditApplicationInstanceHost = (additionalData) => ({
  type: APPLICATION_INSTANCE_HOST_EDIT_ACTIVATE,
  payload: {
    ...additionalData,
  },
});

export const confirmEditApplicationInstanceHost = (rowData) => ({
  type: APPLICATION_INSTANCE_HOST_EDIT_CONFIRM,
  payload: {
    ...rowData,
  },
});

export const cancelEditApplicationInstanceHost = (rowData) => ({
  type: APPLICATION_INSTANCE_HOST_EDIT_CANCEL,
  payload: {
    ...rowData,
  },
});

export const changeEditApplicationInstanceHost = (value, additionalData) => ({
  type: APPLICATION_INSTANCE_HOST_EDIT_CHANGE,
  payload: {
    value,
    ...additionalData,
  },
});

export const openParameterSelectionModal = (additionalData) => ({
  type: APPLICATION_INSTANCE_PARAMETER_SELECTION_MODAL_OPEN,
  payload: {
    ...additionalData,
  },
});

export const closeParameterSelectionModal = (additionalData) => ({
  type: APPLICATION_INSTANCE_PARAMETER_SELECTION_MODAL_CLOSE,
  payload: {
    ...additionalData,
  },
});
