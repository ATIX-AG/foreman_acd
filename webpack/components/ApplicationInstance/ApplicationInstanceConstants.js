export const APPLICATION_INSTANCE_INIT = 'APPLICATION_INSTANCE_INIT';
export const APPLICATION_INSTANCE_CLOSE_ALERT_MODAL = 'APPLICATION_INSTANCE_CLOSE_ALERT_MODAL';
export const APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_REQUEST = 'APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_REQUEST';
export const APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_SUCCESS = 'APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_SUCCESS';
export const APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_FAILURE = 'APPLICATION_INSTANCE_LOAD_APPLICATION_DEFINITION_FAILURE';
export const APPLICATION_INSTANCE_HOST_DELETE = 'APPLICATION_INSTANCE_HOST_DELETE';
export const APPLICATION_INSTANCE_HOST_ADD = 'APPLICATION_INSTANCE_HOST_ADD';
export const APPLICATION_INSTANCE_HOST_EDIT_ACTIVATE = 'APPLICATION_INSTANCE_HOST_EDIT_ACTIVATE';
export const APPLICATION_INSTANCE_HOST_EDIT_CONFIRM = 'APPLICATION_INSTANCE_HOST_EDIT_CONFIRM';
export const APPLICATION_INSTANCE_HOST_EDIT_CHANGE = 'APPLICATION_INSTANCE_HOST_EDIT_CHANGE';
export const APPLICATION_INSTANCE_HOST_EDIT_CANCEL = 'APPLICATION_INSTANCE_HOST_EDIT_CANCEL';
export const APPLICATION_INSTANCE_HOST_EDIT_ERROR = 'APPLICATION_INSTANCE_HOST_EDIT_ERROR';
export const APPLICATION_INSTANCE_FOREMAN_PARAMETER_SELECTION_MODAL_OPEN = 'APPLICATION_INSTANCE_FOREMAN_PARAMETER_SELECTION_MODAL_OPEN';
export const APPLICATION_INSTANCE_FOREMAN_PARAMETER_SELECTION_MODAL_CLOSE = 'APPLICATION_INSTANCE_FOREMAN_PARAMETER_SELECTION_MODAL_CLOSE';
export const APPLICATION_INSTANCE_ANSIBLE_PARAMETER_SELECTION_MODAL_OPEN = 'APPLICATION_INSTANCE_ANSIBLE_PARAMETER_SELECTION_MODAL_OPEN';
export const APPLICATION_INSTANCE_ANSIBLE_PARAMETER_SELECTION_MODAL_CLOSE = 'APPLICATION_INSTANCE_ANSIBLE_PARAMETER_SELECTION_MODAL_CLOSE';
export const APPLICATION_INSTANCE_ADD_EXISTING_HOSTS_MODAL_OPEN = 'APPLICATION_INSTANCE_ADD_EXISTING_HOSTS_MODAL_OPEN';
export const APPLICATION_INSTANCE_ADD_EXISTING_HOSTS_MODAL_CLOSE = 'APPLICATION_INSTANCE_ADD_EXISTING_HOSTS_MODAL_CLOSE';
export const APPLICATION_INSTANCE_CHANGE_PARAMETER_SELECTION_MODE = 'APPLICATION_INSTANCE_CHANGE_PARAMETER_SELECTION_MODE';
