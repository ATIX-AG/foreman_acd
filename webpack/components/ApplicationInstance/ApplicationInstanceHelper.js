
function calculateServiceUsage(hostServiceId, services) {
  const service = services.find(serv => serv['id'] == hostServiceId);
  if ('currentCount' in service) {
    service['currentCount'] += 1;
  } else {
    service['currentCount'] = 1;
  }

  return services;
}

export {
  calculateServiceUsage,
};
