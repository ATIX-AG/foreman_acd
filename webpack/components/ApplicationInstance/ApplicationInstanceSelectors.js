const applicationInstanceConf = state =>
  state.foremanAcd.applicationInstanceConf;

export const selectShowAlertModal = state =>
  applicationInstanceConf(state).showAlertModal;
export const selectAlertModalText = state =>
  applicationInstanceConf(state).alertModalText;
export const selectAlertModalTitle = state =>
  applicationInstanceConf(state).alertModalTitle;
export const selectEditMode = state => applicationInstanceConf(state).editMode;
export const selectAppDefinition = state =>
  applicationInstanceConf(state).appDefinition;
export const selectHosts = state => applicationInstanceConf(state).hosts;
export const selectColumns = state => applicationInstanceConf(state).columns;
export const selectHiddenForemanParameterTypes = state =>
  applicationInstanceConf(state).hiddenForemanParameterTypes;
export const selectServices = state => applicationInstanceConf(state).services;
export const selectParametersData = state =>
  applicationInstanceConf(state).parametersData;
export const selectAnsibleVarsAll = state =>
  applicationInstanceConf(state).ansibleVarsAll;
export const selectParamEditMode = state =>
  applicationInstanceConf(state).paramEditMode;
