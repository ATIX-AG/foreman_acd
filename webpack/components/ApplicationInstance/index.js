import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import './ApplicationInstance.scss';
import ApplicationInstance from './ApplicationInstance';
import * as ApplicationInstanceActions from './ApplicationInstanceActions';

import {
  selectShowAlertModal,
  selectAlertModalText,
  selectAlertModalTitle,
  selectEditMode,
  selectAppDefinition,
  selectHosts,
  selectServices,
  selectColumns,
  selectHiddenForemanParameterTypes,
  selectParametersData,
  selectAnsibleVarsAll,
  selectParamEditMode,
} from './ApplicationInstanceSelectors';

const mapStateToProps = state => ({
  showAlertModal: selectShowAlertModal(state),
  alertModalText: selectAlertModalText(state),
  alertModalTitle: selectAlertModalTitle(state),
  editMode: selectEditMode(state),
  appDefinition: selectAppDefinition(state),
  hosts: selectHosts(state),
  services: selectServices(state),
  columns: selectColumns(state),
  hiddenForemanParameterTypes: selectHiddenForemanParameterTypes(state),
  parametersData: selectParametersData(state),
  ansibleVarsAll: selectAnsibleVarsAll(state),
  paramEditMode: selectParamEditMode(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(ApplicationInstanceActions, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ApplicationInstance);
