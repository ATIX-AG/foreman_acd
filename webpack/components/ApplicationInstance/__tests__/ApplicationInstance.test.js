import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import ApplicationInstance from '../ApplicationInstance';

const noop = () => {};

const fixtures = {
  'should render application instance': {
    data: {
      location: 'Default Location',
      organization: 'Default Organization',
      appDefinition: {},
      hosts: [],
      ansibleVarsAll: [],
    },
    closeAlertModal: noop,
    loadApplicationDefinition: noop,
    initApplicationInstance: noop,
    addApplicationInstanceHost: noop,
  },
};

describe('ApplicationInstance', () =>
  testComponentSnapshotsWithFixtures(ApplicationInstance, fixtures));
