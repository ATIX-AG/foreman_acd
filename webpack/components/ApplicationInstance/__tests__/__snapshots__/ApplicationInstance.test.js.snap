// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationInstance should render application instance 1`] = `
<span>
  <div
    class="service-counter"
  >
    <ServiceCounter
      hostList={Array []}
      serviceList={Array []}
      title="Service counts"
    />
  </div>
  <div>
    <AppDefinitionSelector
      additionalData={
        Object {
          "url": undefined,
        }
      }
      editable={false}
      label="Application Definition"
      onChange={[Function]}
      selectValue=""
      viewText=""
    />
    <p
      style={
        Object {
          "paddingTop": 25,
        }
      }
    >
      <pre>
        App Definition can't be blank
      </pre>
    </p>
  </div>
  <div
    className="form-group"
  >
    <TablePfProvider
      bordered={true}
      className=""
      columns={Array []}
      components={
        Object {
          "body": Object {
            "cell": [Function],
            "row": [Function],
          },
        }
      }
      condensed={false}
      dataTable={true}
      hover={true}
      inlineEdit={true}
      striped={true}
    >
      <Header
        headerRows={
          Array [
            Array [],
          ]
        }
      />
      <Body
        onRow={[Function]}
        rowKey="id"
        rows={Array []}
      />
    </TablePfProvider>
    <AddTableEntry
      disabled={true}
      hidden={false}
      onAddTableEntry={[Function]}
    />
    <span
      style={
        Object {
          "marginLeft": 30,
        }
      }
    >
      Ansible group vars 'all':
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "marginLeft": 10,
          }
        }
      >
        <span
          title="change ansible variables for 'all'"
        >
          A
        </span>
      </Button>
    </span>
  </div>
  <div>
    <ForemanModal
      dialogClassName="param_selection_modal"
      id="AppInstanceForemanParamSelection"
      title="Foreman Parameter specification for Application Instance"
    >
      <Header
        closeButton={false}
      >
        Parameter specification
      </Header>
      <Connect(ParameterSelection)
        data={Object {}}
        editModeCallback={[Function]}
        location="Default Location"
        organization="Default Organization"
        paramType="PARAMETER_SELECTION_PARAM_TYPE_FOREMAN"
      />
      <Footer>
        <div>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="primary"
            disabled={false}
            onClick={[Function]}
          >
            Save
          </Button>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            Cancel
          </Button>
        </div>
      </Footer>
    </ForemanModal>
  </div>
  <div>
    <ForemanModal
      dialogClassName="param_selection_modal"
      id="AppInstanceAnsibleParamSelection"
      title="Ansible group variables for Application Instance"
    >
      <Header
        closeButton={false}
      >
        Parameter specification
      </Header>
      <Connect(ParameterSelection)
        data={Object {}}
        editModeCallback={[Function]}
        location="Default Location"
        organization="Default Organization"
        paramType="PARAMETER_SELECTION_PARAM_TYPE_ANSIBLE"
      />
      <Footer>
        <div>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="primary"
            disabled={false}
            onClick={[Function]}
          >
            Save
          </Button>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            Cancel
          </Button>
        </div>
      </Footer>
    </ForemanModal>
  </div>
  <div />
  <RailsData
    key="applications_instance"
    parameter="hosts"
    value="[]"
    view="app_instance"
  />
  <RailsData
    key="applications_instance"
    parameter="ansible_vars_all"
    value="[]"
    view="app_instance"
  />
</span>
`;
